### PR TITLE
Discover `.pyi` files

### LINKED ISSUE #9097: `.pyi` files are not discovered when running on a folder
### Bug description

```python
# asdf/foo.py
foo = 1

# asdf/bar.pyi
foo = 1
```

### Configuration

_No response_

### Command used

```shell
pylint asdf
```

### Pylint output

```
************* Module foo
asdf\foo.py:1:0: C0104: Disallowed name "foo" (disallowed-name)
asdf\foo.py:1:0: C0104: Disallowed name "foo" (disallowed-name)
```

### Expected behavior

```
************* Module foo
asdf\foo.py:1:0: C0104: Disallowed name "foo" (disallowed-name)
asdf\foo.py:1:0: C0104: Disallowed name "foo" (disallowed-name)
asdf\bar.pyi:1:0: C0104: Disallowed name "foo" (disallowed-name)
asdf\bar.pyi:1:0: C0104: Disallowed name "foo" (disallowed-name)
```

### Pylint version

```shell
pylint 3.0.0a7
astroid 3.0.0
Python 3.11.2 (tags/v3.11.2:878ead1, Feb  7 2023, 16:38:35) [MSC v.1934 64 bit (AMD64)]
```

### OS / Environment

windows 10

### Additional dependencies

_No response_

### PR DESCRIPTION
**Pull request by jacobtylerwalls**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description
Closes #9097

### CODE REVIEW

**Review by DanielNoord — DISMISSED** [review]

**Review by Pierre-Sassoulas — APPROVED** [review]

**Comment by jacobtylerwalls** [comment]
@Pierre-Sassoulas @DanielNoord tried and failed to get `--ignore-patterns` to ignore `.pyi` files. Any regex help?

**Comment by Pierre-Sassoulas** [comment]
I think the ``*`` and dot are inverted in ``*.``, so it's "a ``*`` then one character", instead of "any characters 0 to n times" (``.*``). Maybe we also need an anchor at the end like this ``.*\.pyi$`` ?

File: pylint/lint/pylinter.py [hunk 1]
```diff
@@ -600,7 +600,7 @@ def should_analyze_file(modname: str, path: str, is_argument: bool = False) -> b
         """
         if is_argument:
             return True
-        return path.endswith(".py")
+        return path.endswith(".py") or path.endswith(".pyi")
```
Comment (mbyrnepr2) [inline pylint/lint/pylinter.py:603]:
```suggestion
        return path.endswith(("py", ".pyi"))
```
Either way is fine I just thought I'd throw the idea out there :)

**Comment by jacobtylerwalls** [comment]
Thanks for the regex help. Turned out that was a red herring; I was looking at the wrong part of the logs and not seeing why the job failed. There's an AttributeError if you expand the echo warnings step.

File: pylint/lint/pylinter.py [hunk 2]
```diff
@@ -600,7 +600,7 @@ def should_analyze_file(modname: str, path: str, is_argument: bool = False) -> b
         """
         if is_argument:
             return True
-        return path.endswith(".py")
+        return path.endswith(("py", ".pyi"))
```
Comment (JoakimJoensuu) [inline pylint/lint/pylinter.py:603]:
```suggestion
        return path.endswith((".py", ".pyi"))
```

Comment (jacobtylerwalls) [reply pylint/lint/pylinter.py]: Thanks for catching an error I made in haste!

**Comment by jacobtylerwalls** [comment]
Depends on https://github.com/pylint-dev/astroid/pull/2365

**Review by Pierre-Sassoulas — APPROVED** [review]
I rebased on astroid 3.1.0 upgrade, let's merge when the pipeline is green 🚀

**Comment by jacobtylerwalls** [comment]
Sure. (Let's be sure not to promise full support for .pyi though, we still need astroid 3.2 with https://github.com/pylint-dev/astroid/pull/2375)

**Pull request merged**